src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/CreateFileFromTemplateService.cs

```diff
@@ -0,0 +1,79 @@
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using EnvDTE;
+using EnvDTE80;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    /// This service creates a file from a given file template.
+    /// </summary>
+    [Export(typeof(ICreateFileFromTemplateService))]
+    internal class CreateFileFromTemplateService : ICreateFileFromTemplateService
+    {
+        private IUnconfiguredProjectVsServices _projectVsServices;
+
+        [ImportingConstructor]
+        public CreateFileFromTemplateService(IUnconfiguredProjectVsServices projectVsServices)
+        {
+            Requires.NotNull(projectVsServices, nameof(projectVsServices));
+
+            _projectVsServices = projectVsServices;
+        }
+
+        /// <summary>
+        /// Get the language string to pass to the VS APIs for getting a template.
+        /// </summary>
+        private string GetTemplateLanguage(Project project)
+        {
+            switch (project.CodeModel.Language)
+            {
+                case CodeModelLanguageConstants.vsCMLanguageCSharp:
+                    return "CSharp";
+                case CodeModelLanguageConstants.vsCMLanguageVB:
+                    return "VisualBasic";
+                default:
+                    throw new NotSupportedException("Unrecognized language");
+            }
+        }
+
+        /// <summary>
+        /// Create a file with the given template file and add it to the parent node.
+        /// </summary>
+        /// <param name="templateFile">The name of the template zip file.</param>
+        /// <param name="parentNode">The node to which the new file will be added.</param>
+        /// <param name="specialFileName">The name of the file to be created.</param>
+        /// <returns>true if file is added successfully.</returns>
+        public async Task<bool> CreateFileAsync(string templateFile, IProjectTree parentNode, string specialFileName)
+        {
+            await _projectVsServices.ThreadingService.SwitchToUIThread();
+
+            Project project = _projectVsServices.VsHierarchy.GetProperty<Project>(Shell.VsHierarchyPropID.ExtObject, null);
+            var solution = project?.DTE?.Solution as Solution2;
+            if (solution == null)
+            {
+                return false;
+            }
+
+            string templateFilePath = solution.GetProjectItemTemplate(templateFile, GetTemplateLanguage(project));
+
+            if (templateFilePath != null)
+            {
+                var parentId = parentNode.IsRoot() ? (uint)VSConstants.VSITEMID.Root : (uint)parentNode.Identity;
+                var result = new VSADDRESULT[1];
+                _projectVsServices.VsProject.AddItemWithSpecific(parentId, VSADDITEMOPERATION.VSADDITEMOP_RUNWIZARD, specialFileName, 0, new string[] { templateFilePath }, IntPtr.Zero, 0, Guid.Empty, null, Guid.Empty, result);
```

Comment:
Seems redundant with the VSADDRESULT output unless we want to know exact mode of failure.